### PR TITLE
Fixing bugs in LookupTableCoord slicing

### PR DESCRIPTION
**Pull request by Cadair**
In the process of attempting to address the extra coords part of #334 I noticed there were some issue with the way we handled slicing in the `LookupTableCoord` stack, which would be very problematic when it came to dropping dimensions.

I haven't added any dimension dropping in this PR, axes in the table just become length one, but it hopefully lays the ground work for a follow on PR.

The main change introduced in this PR is to move the generation of the gWCS frame into what was the "delayed model" layer. So the three layers of the stack are now `ExtraCoords > LookupTableCoord > BaseTableCoordinate` where there is a subclass of `BaseTableCoordinate` for each type of table we support. This means all the frame generation and slicing logic are all in a type-specific class and the generation of the frame and model is delayed until the `.wcs` property of the `LookupTableCoord` class is called. This means that all the slicing operations now happen on the actual object the user passes to us, we only convert to a model at the last moment.

### CODE REVIEW

**Review by DanRyanIrish — COMMENTED** [review]
Looks good!  A few minor questions and comments.

**Comment by DanRyanIrish** [comment]
Also, there's no changelog.  I know this isn't changing currently available code, but does alter code described in previous changelogs.  Do we need a changelog here then?

File: ndcube/extra_coords/lookup_table_coord.py [hunk 1]
```diff
@@ -14,6 +14,31 @@
...
+    axes_order = tuple(range(naxes))
+
+    name = None
+    axes_type = "CUSTOM"
+
+    if isinstance(unit, (u.Unit, u.IrreducibleUnit, u.CompositeUnit)):
+        unit = tuple([unit] * naxes)
+
+    if all([u.m.is_equivalent(un) for un in unit]):
+        axes_type = "SPATIAL"
```
[the remark this answers is not in the document]
Comment (Cadair) [reply ndcube/extra_coords/lookup_table_coord.py:30]: Temporal and spectral have their own frames, this is a generic frame distance, pixel or unknown units.

File: ndcube/extra_coords/lookup_table_coord.py [hunk 2]
```diff
@@ -67,28 +92,26 @@ class LookupTableCoord:
...
             if not all(isinstance(lt, type(lt0)) for lt in lookup_tables):
                 raise TypeError("All lookup tables must be the same type")
 
             type_map = {
-                u.Quantity: self._from_quantity,
-                Time: self._from_time,
-                SkyCoord: self._from_skycoord
+                u.Quantity: QuantityTableCoordinate,
+                SkyCoord: SkyCoordTableCoordinate,
+                Time: TimeTableCoordinate,
```
[the remark this answers is not in the document]
Comment (Cadair) [reply ndcube/extra_coords/lookup_table_coord.py:105]: Yes, but I have never implemented that, it should be done but not in this PR.

File: ndcube/extra_coords/lookup_table_coord.py [hunk 3]
```diff
@@ -214,113 +270,201 @@ def _generate_compound_model(cls, *lookup_tables, mesh=True):
...
         components = tuple(getattr(sc.data, comp) for comp in sc.data.components)
         ref_frame = sc.frame.replicate_without_data()
         units = list(c.unit for c in components)
 
-        if names and len(names) != 2:
+        names = self.names
+        if self.names and len(self.names) != 2:
             names = None
 
         # TODO: Currently this limits you to 2D due to gwcs#120
```
[the remark this answers is not in the document]
Comment (Cadair) [reply ndcube/extra_coords/lookup_table_coord.py:405]: yep

File: ndcube/tests/test_lookup_table_coord.py [hunk 4]
```diff
@@ -94,85 +130,54 @@ def test_2d_nout_1_no_mesh():
...
 
-    sub_ltc = ltc[1:3, 1:2]
-    assert sub_ltc.delayed_models[0].lookup_table[0].shape == (2, 1)
-    assert sub_ltc.delayed_models[0].lookup_table[1].shape == (2, 1)
+    # assert isinstance(ltc.wcs, gwcs.WCS)
+    #
+    # sub_ltc = ltc[0:4, 0:5, 0:6]
+    # assert sub_ltc.delayed_models[0].lookup_table[0].shape == (4, )
+    # assert sub_ltc.delayed_models[0].lookup_table[1].shape == (5, )
+    # assert sub_ltc.delayed_models[0].lookup_table[2].shape == (6, )
```
[the remark this answers is not in the document]
Comment (Cadair) [reply ndcube/tests/test_lookup_table_coord.py:160]: It's been open a long time on gWCS, so we are probably going to have to do it ourselves.

File: ndcube/tests/test_lookup_table_coord.py [hunk 5]
```diff
@@ -209,7 +201,8 @@ def test_join_3d():
     assert isinstance(world[0], SkyCoord)
     assert u.allclose(world[1], 0 * u.nm)
 
-    assert u.allclose(ltc.wcs.world_to_pixel(*world), (0, 0, 0))
+    # TODO: Investigate this, something about inverse model
+    # assert u.allclose(ltc.wcs.world_to_pixel(*world), (0, 0, 0))
```
Comment (Cadair) [inline ndcube/tests/test_lookup_table_coord.py:205]: This is a real issue.

**Review by DanRyanIrish — APPROVED** [review]

**Comment by Cadair** [comment]
> Also, there's no changelog. I know this isn't changing currently available code, but does alter code described in previous changelogs. Do we need a changelog here then?

It doesn't alter anything I would consider "user facing" it all behaves the same as before but it just works differently, at least that was my logic for skipping the changelog.

File: ndcube/tests/test_lookup_table_coord.py [hunk 6]
```diff
@@ -5,41 +5,89 @@
...
+    return TimeTableCoordinate(data)
+
+
+@pytest.fixture
+def lut_1d_wave():
+    # TODO: Make this into a SpectralCoord object
+    return QuantityTableCoordinate(range(10) * u.nm)
+
+
+# def test_exceptions(lut_1d_distance):
```
Comment (Cadair) [inline ndcube/tests/test_lookup_table_coord.py:73]: fix me

File: ndcube/extra_coords/extra_coords.py [hunk 7]
```diff
@@ -298,10 +312,13 @@ def _getitem_lookup_tables(self, item):
...
             lut_axes = (lut_axis,) if not isinstance(lut_axis, tuple) else lut_axis
             lut_slice = tuple(item[i] for i in lut_axes) if isinstance(item, tuple) else item
+            if isinstance(lut_slice, tuple) and len(lut_slice) == 1:
+                lut_slice = lut_slice[0]
 
             sliced_lut = lut[lut_slice]
-            if sliced_lut:
+            if not isinstance(lut_slice, Integral):
                 new_lookup_tables.add((lut_axis, sliced_lut))
+            # TODO: Handle dropped table here
```
Comment (DanRyanIrish) [inline ndcube/extra_coords/extra_coords.py:321]: Is this to be fixed in this PR or is this where a new PR should address #349?

File: ndcube/extra_coords/lookup_table_coord.py [hunk 8]
```diff
@@ -9,318 +9,431 @@
...
-                SkyCoord: self._from_skycoord
-            }
-            delayed_model, frame = type_map[type(lt0)](lookup_tables,
-                                                       mesh=mesh,
-                                                       names=names,
-                                                       physical_types=physical_types)
-            self.delayed_models = [delayed_model]
-            self.frames = [frame]
+    ndim = lookup_table.ndim
+    TabularND = tabular_model(ndim, name=f"Tabular{ndim}D")
```
Comment (DanRyanIrish) [inline ndcube/extra_coords/lookup_table_coord.py:49]: Is this returning an class or an instance of a class?  If the latter, perhaps it shouldn't be camelcase.

**Comment by DanRyanIrish** [comment]
This updated implementation looks good.  Barring further work on the tests, am happy if you feel the core code is done.

File: ndcube/extra_coords/lookup_table_coord.py (hunk 8 above)
Comment (Cadair) [reply ndcube/extra_coords/lookup_table_coord.py:49]: It returns a class.

File: ndcube/extra_coords/extra_coords.py (hunk 7 above)
Comment (Cadair) [reply ndcube/extra_coords/extra_coords.py:321]: Probably a follow on one.

**Comment by rosteen** [comment]
Looks good to me.

**Pull request merged**